### PR TITLE
win: allow for already bound/connected socket in uv_tcp_open

### PR DESCRIPTION
**Pull request by Neverous**
On Unix you can pass bound and even connected socket to uv_tcp_open
and it will work as expected: you can run uv_listen or other io
functions (uv_read, uv_write) on it without problems.
On windows on the other hand the function always assumes to have clean
socket and for example uv_listen will try to bind it again, and
uv_read/write will return with errors about unreadable streams
(basically invalid internal flags).

To check if socket is already bound uv_tcp_getsockname is called which
on windows returns error when socket is unbound. To further
differentiate connected one from just bound, uv_tcp_getpeername also
returns error but when target socket is not connected.

Loosely based on (dead?) https://github.com/libuv/libuv/pull/1150 . This builds upon it and adds support for connected sockets.

### CODE REVIEW

File: src/win/tcp.c [hunk 1]
```diff
@@ -1446,6 +1446,8 @@ int uv_tcp_open(uv_tcp_t* handle, uv_os_sock_t sock) {
   WSAPROTOCOL_INFOW protocol_info;
   int opt_len;
   int err;
+  struct sockaddr_storage saddr;
+  int saddr_len = sizeof(saddr);
```
Comment (bnoordhuis) [inline src/win/tcp.c]: Minor stylistic issue: can you change this to just `int saddr_len;` and do the assignment a few lines below?

File: src/win/tcp.c [hunk 2]
```diff
@@ -1466,6 +1468,17 @@ int uv_tcp_open(uv_tcp_t* handle, uv_os_sock_t sock) {
     return uv_translate_sys_error(err);
   }
 
+  /* Support already active socket. */
+  if(!uv_tcp_getsockname(handle, (struct sockaddr*) &saddr, &saddr_len)) {
```
Comment (bnoordhuis) [inline src/win/tcp.c]: Space before `(`.  Likewise below.

**Comment by bnoordhuis** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/372/

**Comment by Neverous** [comment]
fixed style issues, and also added resetting saddr_len before getpeername because getsockname can change it.

**Comment by bzoz** [comment]
CI after fixup: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/373/

**Comment by bnoordhuis** [comment]
Thanks Maciej, landed in 6827fa3.

**Comment by saghul** [comment]
Sorry I'm late, I'd just wanted to say this is awesome. Thanks @Neverous!

**Pull request closed without merging**